src/Network/Wai/SAML2/Validation.hs

```diff
@@ -185,10 +186,21 @@ validateResponse cfg responseData = runExceptT $ do
     -- validate that the assertion is valid at this point in time
     let Conditions{..} = assertionConditions assertion
 
-    if (now < conditionsNotBefore || now >= conditionsNotOnOrAfter) &&
-        not (saml2DisableTimeValidation cfg)
-    then throwError NotValid
-    else pure ()
+    -- Reference [NotBefore and NotOnOrAfter]
+    when ((now < conditionsNotBefore || now >= conditionsNotOnOrAfter) &&
+           not (saml2DisableTimeValidation cfg))
+          $ throwError NotValid
+
+    -- Reference [AudienceRestriction]
+    -- Note [Validating AudienceRestrictions]
+    case saml2Audiences cfg of
+        -- Check disabled
+        [] -> pure ()
+        ourAudiences ->
+           forM_ conditionsAudienceRestrictions $
+              \(AudienceRestriction audiences) ->
+                 unless (any (`elem` ourAudiences) audiences)
+                   $ throwError (AudienceMismatch audiences)
```

Comment:
It might be worthwhile to annotate this part of the code with an explanation of the intended logic. It seems that the specification suggests that different `Audience` URIs within an `AudienceRestriction` form a disjunction and multiple `AudienceRestrictions` form a conjunction. This logic seems to be implemented here as well (i.e. an error is thrown if, for any `AudienceRestriction`, there isn't a match with one of the `Audience`s). For clarity, this should be documented here inline.